### ISSUE TITLE
Query compiler gives dangerous advice

### ISSUE THREAD

**Issue by dthul** [elixir-lang/ecto#1217]
When joining associations and trying to specify an `on` clause the query compiler throws an error and tells the user to move it into a `where` clause instead.
For example when I want to load the comments of a post that were written in the last 12 hours:

```elixir
from p in Posts,
  left_join: c in assoc(p, :comments),
    on: c.creation_time > fragment("now() - interval '12 hours'"),
  preload: [comments: c]
```
The query compiler will tell me:
```
== Compilation error on file web/controllers/PostsController.ex ==
** (Ecto.Query.CompileError) cannot specify `on` on `left_join` when using association join, add extra clauses with `where` instead
    lib/ecto/query.ex:331: Ecto.Query.from/5
    expanding macro: Ecto.Query.from/2
    web/controllers/PostsController.ex:13: App.PostsController.index/2
    (elixir) lib/kernel/parallel_compiler.ex:100: anonymous fn/4 in Kernel.ParallelCompiler.spawn_compilers/8
```
An unsuspecting user might follow that advice and rewrite the query like so:
```elixir
from p in Posts,
  left_join: c in assoc(p, :comments),
  where: c.creation_time > fragment("now() - interval '12 hours'"),
  preload: [comments: c]
```
This compiles but is not at all equivalent to the original version and intention of the user. Even worse the newly introduced bug will only show when there have been no comments in the last 12 hours. The left join was used to make sure that even if a post doesn't have comments it will show up in the query result. But since the condition has been moved to the `where` clause (as suggested by the query compiler) posts that had no comments in the last 12 hours will be filtered away.

That's why I suggest to remove this "hint" or at least reword it to make explicit that it only works for specific kinds of joins (like an inner join).

**Comment by josevalim** [comment]
@dthul excellent point. I will just allow both to be supported and measure the clauses with and.
